segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+     * </li>
+     * <li>
+     * <li>
+     * Defrag operation will respect max rolling size and will not create chunks greater than that size.
+     * </li>
+     * </ul>
+     *
+     * <div>
+     * What controls whether we invoke native concat or simulate through appends?
+     * There are a few different capabilities that ChunkStorageProvider needs to provide.
+     * <ul>
+     * <li>Does ChunkStorageProvider support appending to existing chunks? For vanilla S3 compatible this would return false.
+     * This is indicated by supportsAppend.</li>
+     * <li>Does ChunkStorageProvider support for concatenating chunks natively? This is indicated by supportsConcat.
+     * If this is true then native concat operation concat will be invoked otherwise concatWithAppend is invoked.</li>
+     * <li>There are some obvious constraints - For ChunkStorageProvider support any concat functionality it must support either append or native concat.</li>
+     * <li>Also when ChunkStorageProvider supports both native and append, ChunkStorageManager will invoke appropriate method
+     * depending on size of target and source chunks. (Eg. ECS)</li>
+     * </ul>
+     * </div>
+     * <li>
+     * What controls defrag?
+     * There are two additional parameters that control when native concat
+     * <li>minSizeLimitForNativeConcat : Size of chunk in bytes above which it is no longer considered a small object. For small source objects, concatWithAppend is used instead of using concat. (For really small txn it is rather efficient to use append than MPU).</li>
+     * <li>maxSizeLimitForNativeConcat: Size of chunk in bytes above which it is no longer considered for concat. (Eg S3 might have max limit on chunk size).</li>
+     * In short there is a size beyond which using append is not advisable. Conversely there is a size below which native concat is not efficient.(minSizeLimitForNativeConcat )
+     * Then there is limit which concating does not make sense maxSizeLimitForNativeConcat
+     * </li>
+     * <li>
+     * What is the defrag algorithm
+     * <pre>
+     * While(segment.hasConcatableChunks()){
+     *     Set<List<Chunk>> s = FindConsecutiveConcatableChunks();
+     *     For (List<chunk> list : s){
+     *        ConcatChunks (list);
+     *     }
+     * }
+     * </pre>
+     * </li>
+     * </ul>
+     *
+     * @param txn             Active {@link MetadataTransaction}.
+     * @param segmentMetadata {@link SegmentMetadata} for the segment to defrag.
+     * @param startChunkName  Name of the first chunk to start defragmentation.
+     * @param lastChunkName   Name of the last chunk before which to stop defragmentation. (last chunk is not concatenated).
+     * @throws Exception In case of any errors.
+     */
+    private void defrag(MetadataTransaction txn, SegmentMetadata segmentMetadata, String startChunkName, String lastChunkName) throws Exception {
```

Comment:
If there is a background task doing defragmentation, then do we still need a metadata transaction for context?

Reply:
No background task today. The defrag is called inline at the end of concat.
I do plan to add background task in near future. (May be 0.9 time frame)